src/sentry/api/serializers/models/project.py

```diff
@@ -187,8 +187,33 @@ def _get_features_for_projects(all_projects, user):
             projects_by_org[project.organization].append(project)
 
         features_by_project = defaultdict(list)
+        project_features = [
+            feature
+            for feature in features.all(feature_type=ProjectFeature).keys()
+            if feature.startswith(_PROJECT_SCOPE_PREFIX)
+        ]
+
+        batch_checked = set()
+        for (organization, projects) in projects_by_org.items():
+            batch_features = features.batch_has(
+                project_features, actor=user, projects=projects, organization=organization
+            )
+
+            # batch_has has found some features
+            if batch_features:
+                for project in projects:
+                    for feature_name, active in batch_features.get(
+                        "project:{}".format(project.id), {}
+                    ):
+                        if active:
+                            features_by_project[project].append(
+                                feature_name[len(_PROJECT_SCOPE_PREFIX) :]
+                            )
+
+                        batch_checked.add(feature_name)
+
         for feature_name in features.all(feature_type=ProjectFeature).keys():
```

Comment:
Should you use the same pattern you used for org features? In organization features the feature manager is only queried once and as features are found in the batch response they are removed from features to be checked individually.

Reply:
So for the project serializer this I can't remove from the `project_features` list, the serializer code is setup in such a way that there could be more than [one org's worth of projects here](https://github.com/getsentry/sentry/pull/21180/files#diff-d42054a9794cab517fb89e22b8b50de14d73cdde13ed7cfe9fd6481dc3f9fb21R185-R187).
This means that `batch_has` needs to be checked per potential org (and its projects)
But these flags that are `batch_checked` are [removed from future checks here](https://github.com/getsentry/sentry/pull/21180/files#diff-d42054a9794cab517fb89e22b8b50de14d73cdde13ed7cfe9fd6481dc3f9fb21R216-R217)